README.md
# Stand-Alone-Self-Attention-in-Vision-Models-

- Reproduced and confirmed the results from above NIPS 2019 research paper for course project.
- Implemented ResNet50 convolution network with self-attention layers as per the guidelines in the paper.
- Analyzed effect of spatial extent, positional embedding, and attention type on ResNet50 attention model.
- Implemented SSD object detection model by replacing set of convolution layers with self-attention layers.
- Evaluated above SSD model for performance change in terms of Accuracy (4.4% less), number of model parameters (5.6% less), and Flops count (very less).


misc/history
    1  20180904_180723  ls
    2  20180904_180828  sftp ashinmarin@huckleberry1.arc.vt.edu
    3  20180920_114342  $ module load Anaconda/5.1.0
    4  20180920_114432  source activate pytorch_env
    5  20180920_114443  python
    6  20180920_120013  exit() cd..
    7  20180925_111304  module load Anaconda/5.1.0
    8  20180925_111313  cp -r /groups/srijithr_shared/pytorch_env ~/.conda/envs/
    9  20180925_111338  ls
   10  20180925_111547  conda create -n test
   11  20180925_111611  cp -r /groups/srijithr_shared/pytorch_env ~/.conda/envs/
   12  20180925_114140  conda env list
   13  20180925_114510  interact -l walltime=01:00:00
   14  20180926_123435  #!/bin/bash
   15  20180926_123443  #SBATCH -J hello-world
   16  20180926_123454  #SBATCH -p normal_q
   17  20180926_123509  #SBATCH -N 1
   18  20180926_123518  #SBATCH -t 10:00
   19  20180926_123527  #SBATCH --mem=500G
   20  20180926_123536  #SBATCH --gres=gpu:pascal:4
   21  20180926_123544  echo "hello world"
   22  20180926_123716  ashinmarin@hulogin1:~/Slurm$ sbatch hello.sh
   23  20180926_123738  ~/Slurm$ sbatch hello.sh
   24  20180926_123740  ls
   25  20180926_123800  ~/Slurm$ squeue -u mcclurej
   26  20180926_123815  ~/Slurm$ squeue -u ashinmarin
   27  20180926_123857  salloc -N1 -t 10:00
   28  20180926_123952  jobid
   29  20180926_124007  echo "hello world"
   30  20180926_124023  salloc -n1 -t 10:00 --mem-per-cpu=8G --gres=gpu:pascal:1
   31  20180926_124411  salloc -N1 -t 10:00
   32  20180926_125309  scp temp.txt ashinmarin@huckleberry-1.arc.vt.edu:myfolder/blueridge.txt
   33  20180926_125357  sftp ashinmarin@huckleberry1.arc.vt.edu
   34  20180928_150052  #!/bin/bash
   35  20180928_150233  #SBATCH -J hello-world
   36  20180928_150241  #SBATCH -p normal_q
   37  20180928_150248  #SBATCH -N 1
   38  20180928_150256  #SBATCH -t 10:00
   39  20180928_150304  #SBATCH --mem=500G
   40  20180928_150310  #SBATCH --gres=gpu:pascal:4
   41  20180928_150318  echo "hello world"
   42  20180928_150331  salloc -N1 -t 10:00
   43  20181015_151110  ls
   44  20181015_151335  sftp ashinmarin@cascades1.arc.vt.edu
   45  20181015_151938  exit
   46  20181015_165356  ls
   47  20181015_165359  cd work
   48  20181015_165415  lls
   49  20181015_165436  lpwd
   50  20181015_165749  lcd home
   51  20181015_170236  sftp ashinmarin@cascades1.arc.vt.edu
   52  20181015_172249  module load python
   53  20181015_172305  python prep_data.py
   54  20181015_172317  ls
   55  20181015_172331  cd text-summarization-tensorflow-master
   56  20181015_172339  ls
   57  20181015_172352  python prep_data.py
   58  20181015_172429  pip install wget
   59  20181015_172529  sudo pip install wget
   60  20181015_172713  pip install wget
   61  20181015_172843  pip install --upgrade pip
   62  20181015_173234  export PYTHONUSERBASE=/home/<ashinmarin>/cascades/python
   63  20181015_173245  exit
   64  20181015_173659  ls
   65  20181015_173722  export PYTHONUSERBASE=/home/<ashinmarin>/cascades/python
   66  20181015_173736  export PYTHONUSERBASE=/home/ashinmarin/cascades/python
   67  20181015_173742  ls
   68  20181015_173755  module load gcc/5.2.0 python
   69  20181015_173820  pip install wget
   70  20181015_173833  module load gcc/5.2.0 python atlas numpy scipy
   71  20181015_173847  pip install wget
   72  20181015_173859  pip install --user wget
   73  20181015_173931  easy_install --user wget
   74  20181015_173944  module load gcc/5.2.0 python
   75  20181015_173955  easy_install --user wget
   76  20181015_174038  export PYTHONUSERBASE=/home/ashinmarin/cascades/python
   77  20181015_174053  cd PYTHONUSERBASE
   78  20181015_174112  module load gcc/5.2.0 python
   79  20181015_174124  easy_install --user wget
   80  20181015_174138  easy_install --user matplotlib
   81  20181015_174212  ls
   82  20181018_112802  module load Anaconda
   83  20181018_112832  easy_install --user wget
   84  20181018_112934  cd text-summarization-tensorflow-master
   85  20181018_112936  ls
   86  20181018_112941  ls
   87  20181018_114252  python prep_data.py
   88  20181018_114323  python train.py
   89  20181018_114349  ls
   90  20181018_114747  python test.py
   91  20181018_114759  python train.py
   92  20181018_114825  easy_install --user gensim
   93  20181018_115103  easy_install --user nltk
   94  20181018_115118  cd text-summarization-tensorflow-master
   95  20181018_115122  ls
   96  20181018_115130  python train.py
   97  20181018_115404  module load Anaconda3
   98  20181018_115408  ls
   99  20181018_115442  easy_install --user nltk
  100  20181018_115549  python text.py
  101  20181018_115555  python test.py
  102  20181018_115608  easy_install --user tensorflow
  103  20181018_115635  easy_install --user gensim
  104  20181018_120321  pip install tensorflow
  105  20181018_120412  python train.py
  106  20181018_172047  ls
  107  20181018_172140  qsub ./Cascades_TensorFlow_Test.sh
  108  20181018_172229  python Cascades_TensorFlow_Test.py
  109  20181018_172334  module load Anaconda3
  110  20181018_172533  wget https://repo.anaconda.com/archive/Anaconda3-5.2.0-Linux-x86_64.sh
  111  20181018_172600  bash Anaconda3-5.2.0-Linux-x86_64.sh
  112  20181018_174305  conda install pytorch=0.3.1 cuda90 -c pytorch
  113  20181018_174446  ls
  114  20181018_174504  ./anaconda3/bin/conda 
  115  20181018_174531  ./anaconda3/bin/conda install pytorch=0.3.1 cuda90 -c pytorch
  116  20181018_175037  ls
  117  20181018_175104  qsub Cascades_TensorFlow_Test.sh
  118  20181018_175129  conda install tensorflow-gpu
  119  20181018_180207  qsub Cascades_TensorFlow_Test.sh
  120  20181018_180219  ls
  121  20181018_180311  python Cascades_TensorFlow_Test.py
  122  20181018_180431  vim Cascades_TensorFlow_Test.sh.o297604
  123  20181018_180444  ls -l
  124  20181018_180452  vim Cascades_TensorFlow_Test.sh.e297604
  125  20181018_180750  qsub Cascades_TensorFlow_Test.sh
  126  20181018_180810  ls
  127  20181018_180908  ls -l

cifar100/conv_stem/attn50_cifar100.py

#Library Utilities
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.nn.init as init
import torch.optim as optim
from torchvision import datasets, transforms
from torchvision import models

import math
import os
from tqdm import tqdm
import shutil


# In[37]:


#Attention Layers
#Type 1
#Convolution Attention
class AttentionStem(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, groups=1, m=4, bias=False):
        super(AttentionStem, self).__init__()
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.groups = groups
        self.m = m

        assert self.out_channels % self.groups == 0, "out_channels should be divided by groups. (example: out_channels: 40, groups: 4)"

        self.emb_a = nn.Parameter(torch.randn(out_channels // groups, kernel_size), requires_grad=True)
        self.emb_b = nn.Parameter(torch.randn(out_channels // groups, kernel_size), requires_grad=True)
        self.emb_mix = nn.Parameter(torch.randn(m, out_channels // groups), requires_grad=True)

        self.key_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias)
        self.query_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias)
        self.value_conv = nn.ModuleList([nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias) for _ in range(m)])

        self.reset_parameters()

    def forward(self, x):
        batch, channels, height, width = x.size()

        padded_x = F.pad(x, [self.padding, self.padding, self.padding, self.padding])

        q_out = self.query_conv(x)
        k_out = self.key_conv(padded_x)
        v_out = torch.stack([self.value_conv[_](padded_x) for _ in range(self.m)], dim=0)

        k_out = k_out.unfold(2, self.kernel_size, self.stride).unfold(3, self.kernel_size, self.stride)
        v_out = v_out.unfold(3, self.kernel_size, self.stride).unfold(4, self.kernel_size, self.stride)

        k_out = k_out[:, :, :height, :width, :, :]
        v_out = v_out[:, :, :, :height, :width, :, :]

        emb_logit_a = torch.einsum('mc,ca->ma', self.emb_mix, self.emb_a)
        emb_logit_b = torch.einsum('mc,cb->mb', self.emb_mix, self.emb_b)
        emb = emb_logit_a.unsqueeze(2) + emb_logit_b.unsqueeze(1)
        emb = F.softmax(emb.view(self.m, -1), dim=0).view(self.m, 1, 1, 1, 1, self.kernel_size, self.kernel_size)

        v_out = emb * v_out

        k_out = k_out.contiguous().view(batch, self.groups, self.out_channels // self.groups, height, width, -1)
        v_out = v_out.contiguous().view(self.m, batch, self.groups, self.out_channels // self.groups, height, width, -1)
        v_out = torch.sum(v_out, dim=0).view(batch, self.groups, self.out_channels // self.groups, height, width, -1)

        q_out = q_out.view(batch, self.groups, self.out_channels // self.groups, height, width, 1)

        out = q_out * k_out
        out = F.softmax(out, dim=-1)
        out = torch.einsum('bnchwk,bnchwk->bnchw', out, v_out).view(batch, -1, height, width)

        return out

    def reset_parameters(self):
        init.kaiming_normal_(self.key_conv.weight, mode='fan_out', nonlinearity='relu')
        init.kaiming_normal_(self.query_conv.weight, mode='fan_out', nonlinearity='relu')
        for _ in self.value_conv:
            init.kaiming_normal_(_.weight, mode='fan_out', nonlinearity='relu')

        init.normal_(self.emb_a, 0, 1)
        init.normal_(self.emb_b, 0, 1)
        init.normal_(self.emb_mix, 0, 1)

class AttentionConv(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, groups=1, bias=False):
        super(AttentionConv, self).__init__()
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.groups = groups

        assert self.out_channels % self.groups == 0, "out_channels should be divided by groups. (example: out_channels: 40, groups: 4)"

        self.rel_h = nn.Parameter(torch.randn(out_channels // 2, 1, 1, kernel_size, 1), requires_grad=True)
        self.rel_w = nn.Parameter(torch.randn(out_channels // 2, 1, 1, 1, kernel_size), requires_grad=True)

        self.key_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias)
        self.query_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias)
        self.value_conv = nn.Conv2d(in_channels, out_channels, kernel_size=1, bias=bias)

        self.reset_parameters()

    def forward(self, x):
        batch, channels, height, width = x.size()

        padded_x = F.pad(x, [self.padding, self.padding, self.padding, self.padding])
        q_out = self.query_conv(x)
        k_out = self.key_conv(padded_x)
        v_out = self.value_conv(padded_x)

        k_out = k_out.unfold(2, self.kernel_size, self.stride).unfold(3, self.kernel_size, self.stride)
        v_out = v_out.unfold(2, self.kernel_size, self.stride).unfold(3, self.kernel_size, self.stride)

        #Correction 1
        #v_out_h, v_out_w = v_out.split(self.out_channels // 2, dim=1)
        #v_out = torch.cat((v_out_h + self.rel_h, v_out_w + self.rel_w), dim=1)
        k_out_h, k_out_w = k_out.split(self.out_channels // 2, dim=1)
        k_out = torch.cat((k_out_h + self.rel_h, k_out_w + self.rel_w), dim=1)

        k_out = k_out.contiguous().view(batch, self.groups, self.out_channels // self.groups, height, width, -1)
        v_out = v_out.contiguous().view(batch, self.groups, self.out_channels // self.groups, height, width, -1)

        q_out = q_out.view(batch, self.groups, self.out_channels // self.groups, height, width, 1)

        #Correction 2
        out = q_out * k_out
        #out = (q_out * k_out).sum(dim=2)
        
        out = F.softmax(out, dim=-1)
        out = torch.einsum('bnchwk,bnchwk -> bnchw', out, v_out).view(batch, -1, height, width)

        return out

    def reset_parameters(self):
        init.kaiming_normal_(self.key_conv.weight, mode='fan_out', nonlinearity='relu')
        init.kaiming_normal_(self.value_conv.weight, mode='fan_out', nonlinearity='relu')
        init.kaiming_normal_(self.query_conv.weight, mode='fan_out', nonlinearity='relu')

        init.normal_(self.rel_h, 0, 1)
        init.normal_(self.rel_w, 0, 1)

# temp = torch.randn((2, 3, 32, 32))
# conv = AttentionConv(3, 16, kernel_size=3, padding=1)
# print(conv(temp).size())


# In[38]:


class Bottleneck(nn.Module):
    expansion = 4

    def __init__(self, in_channels, out_channels, stride=1, groups=1, base_width=64):
        super(Bottleneck, self).__init__()
        self.stride = stride
        width = int(out_channels * (base_width / 64.)) * groups

        self.conv1 = nn.Sequential(
            nn.Conv2d(in_channels, width, kernel_size=1, bias=False),
            nn.BatchNorm2d(width),
            nn.ReLU(),
        )
        self.conv2 = nn.Sequential(
            AttentionConv(width, width, kernel_size=7, padding=3, groups=8),
            nn.BatchNorm2d(width),
            nn.ReLU(),
        )
        self.conv3 = nn.Sequential(
            nn.Conv2d(width, self.expansion * out_channels, kernel_size=1, bias=False),
            nn.BatchNorm2d(self.expansion * out_channels),
        )

        self.shortcut = nn.Sequential()
        if stride != 1 or in_channels != self.expansion * out_channels:
            self.shortcut = nn.Sequential(
                nn.Conv2d(in_channels, self.expansion * out_channels, kernel_size=1, stride=stride, bias=False),
                nn.BatchNorm2d(self.expansion * out_channels)
            )

    def forward(self, x):
        out = self.conv1(x)
        out = self.conv2(out)
        out = self.conv3(out)
        if self.stride >= 2:
            out = F.avg_pool2d(out, (self.stride, self.stride))

        out += self.shortcut(x)
        out = F.relu(out)

        return out


# In[39]:


class Model(nn.Module):
    def __init__(self, block, num_blocks, num_classes=1000, stem=False):
        super(Model, self).__init__()
        self.in_places = 64

        if stem:
            self.init = nn.Sequential(
                # CIFAR10
                AttentionStem(in_channels=3, out_channels=64, kernel_size=4, stride=1, padding=2, groups=1),
                nn.BatchNorm2d(64),
                nn.ReLU(),

                # For ImageNet
                # AttentionStem(in_channels=3, out_channels=64, kernel_size=4, stride=1, padding=2, groups=1),
                # nn.BatchNorm2d(64),
                # nn.ReLU(),
                # nn.MaxPool2d(4, 4)
            )
        else:
            self.init = nn.Sequential(
                # CIFAR10
                nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False),
                nn.BatchNorm2d(64),
                nn.ReLU(),

                # For ImageNet
                # nn.Conv2d(3, 64, kernel_size=7, stride=2, padding=3, bias=False),
                # nn.BatchNorm2d(64),
                # nn.ReLU(),
                # nn.MaxPool2d(kernel_size=3, stride=2, padding=1),
            )

        self.layer1 = self._make_layer(block, 64, num_blocks[0], stride=1)
        self.layer2 = self._make_layer(block, 128, num_blocks[1], stride=2)
        self.layer3 = self._make_layer(block, 256, num_blocks[2], stride=2)
        self.layer4 = self._make_layer(block, 512, num_blocks[3], stride=2)
        self.dense = nn.Linear(512 * block.expansion, num_classes)

    def _make_layer(self, block, planes, num_blocks, stride):
        strides = [stride] + [1] * (num_blocks - 1)
        layers = []
        for stride in strides:
            layers.append(block(self.in_places, planes, stride))
            self.in_places = planes * block.expansion
        return nn.Sequential(*layers)

    def forward(self, x):
        out = self.init(x)
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out)
        out = F.avg_pool2d(out, 4)
        out = out.view(out.size(0), -1)
        out = self.dense(out)

        return out


def ATNnet26(num_classes=10, stem=False):
    return Model(Bottleneck, [1, 2, 4, 1], num_classes=num_classes, stem=stem)


def ATNnet38(num_classes=10, stem=False):
    return Model(Bottleneck, [2, 3, 5, 2], num_classes=num_classes, stem=stem)


def ATNnet50(num_classes=10, stem=False):
    return Model(Bottleneck, [3, 4, 6, 3], num_classes=num_classes, stem=stem)


# In[40]:


def get_model_parameters(model):
    total_parameters = 0
    for layer in list(model.parameters()):
        layer_parameter = 1
        for l in list(layer.size()):
            layer_parameter *= l
        total_parameters += layer_parameter
    return total_parameters


# In[41]:


EPOCHS=100
INIT_LR=1e-1

num_workers=8
img_size=32
batch_size=25
momentum=0.9
weight_decay=5e-4
num_classes = 100
model_name='Resnet50'

STEM_CONFIG=False
print_interval=500

# In[42]:


def adjust_learning_rate(optimizer, epoch):
    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
    lr = INIT_LR * (0.1 ** (epoch // 30))
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr


def train(model, train_loader, optimizer, criterion, epoch):
    model.train()

    train_acc = 0.0
    step = 0
    for data, target in train_loader:
        adjust_learning_rate(optimizer, epoch)
        
        data, target = data.cuda(), target.cuda()

        optimizer.zero_grad()
        output = model(data)
        loss = criterion(output, target)
        loss.backward()
        optimizer.step()

        y_pred = output.data.max(1)[1]

        acc = float(y_pred.eq(target.data).sum()) / len(data) * 100.
        train_acc += acc
        step += 1
        if step % print_interval == 0:
            print("[Epoch {0:4d}] Loss: {1:2.3f} Acc: {2:.3f}%".format(epoch, loss.data, acc), end='')
            #logger.info("[Epoch {0:4d}] Loss: {1:2.3f} Acc: {2:.3f}%".format(epochs, loss.data, acc))
            for param_group in optimizer.param_groups:
                print(",  Current learning rate is: {}".format(param_group['lr']))
                #logger.info("Current learning rate is: {}".format(param_group['lr']))


def eval(model, test_loader):
    print('evaluation ...')
    model.eval()
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.cuda(), target.cuda()
            output = model(data)
            prediction = output.data.max(1)[1]
            correct += prediction.eq(target.data).sum()

    acc = 100. * float(correct) / len(test_loader.dataset)
    print('Test acc: {0:.2f}'.format(acc))
    return acc


# In[ ]:





# In[43]:


# Data loading code
transform_train = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=(0.5071, 0.4865, 0.4409),
        std=(0.2673, 0.2564, 0.2762)
    ),
])
transform_test = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(
        mean=(0.5071, 0.4865, 0.4409),
        std=(0.2673, 0.2564, 0.2762)
    ),
])


train_loader = torch.utils.data.DataLoader(
    datasets.CIFAR100('data', train=True, download=True, transform=transform_train),
    batch_size=batch_size,
    shuffle=True,
    num_workers=num_workers
)

test_loader = torch.utils.data.DataLoader(
    datasets.CIFAR100('data', train=False, download=True, transform=transform_test),
    batch_size=batch_size,
    shuffle=False,
    num_workers=num_workers
)


# In[44]:


def save_checkpoint(state, is_best, filename):
    file_path = os.path.join('./checkpoint', filename)
    torch.save(state, file_path)
    best_file_path = os.path.join('./checkpoint', 'best_' + filename)
    if is_best:
        print('best Model Saving ...')
        shutil.copyfile(file_path, best_file_path)


# In[45]:


print('img_size: {}, num_classes: {}, stem: {}'.format(img_size, num_classes, True))

if model_name=='ATNnet50': 
    print("Running ATNnet50")
    if STEM_CONFIG:
        print("Stem is Attention")		
    else:
        print("Stem is Convolution")		
    model = ATNnet50(num_classes=num_classes, stem=STEM_CONFIG)
else:
    print("Running Resnet50")
    model = models.resnet50()

print("Number of model parameters: ", get_model_parameters(model))

start_epoch = 1
best_acc = 0.0

if torch.cuda.device_count() > 1:
    model = nn.DataParallel(model)
model = model.cuda()



criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=INIT_LR, momentum=momentum, weight_decay=weight_decay)

for epoch in range(start_epoch, EPOCHS + 1):
    train(model, train_loader, optimizer, criterion, epoch)
    eval_acc = eval(model, test_loader)

    is_best = eval_acc > best_acc
    best_acc = max(eval_acc, best_acc)

    if is_best:
        print("printing best accuracy so far{}".format(best_acc))
        
    if not os.path.isdir('checkpoint'):
        os.mkdir('checkpoint')
    filename = 'model_ckpt.tar'
    print('filename :: ', filename)

    parameters = get_model_parameters(model)

    if torch.cuda.device_count() > 1:
        save_checkpoint({
            'epoch': epoch,
            'arch': model_name,
            'state_dict': model.module.state_dict(),
            'best_acc': best_acc,
            'optimizer': optimizer.state_dict(),
            'parameters': parameters,
        }, is_best, filename)
    else:
        save_checkpoint({
            'epoch': epoch,
            'arch': model_name,
            'state_dict': model.state_dict(),
            'best_acc': best_acc,
            'optimizer': optimizer.state_dict(),
            'parameters': parameters,
        }, is_best, filename)
